okio/src/main/java/okio/BufferedSink.kt

```diff
@@ -0,0 +1,358 @@
+/*
+ * Copyright (C) 2014 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okio
+
+import java.io.IOException
+import java.io.OutputStream
+import java.nio.channels.WritableByteChannel
+import java.nio.charset.Charset
+
+/**
+ * A sink that keeps a buffer internally so that callers can do small writes
+ * without a performance penalty.
+ */
+interface BufferedSink : Sink, WritableByteChannel {
+  /** Returns this sink's internal buffer. */
+  fun buffer(): Buffer
+
+  @Throws(IOException::class)
+  fun write(byteString: ByteString): BufferedSink
+
+  /**
+   * Like [OutputStream.write], this writes a complete byte array to
+   * this sink.
+   */
+  @Throws(IOException::class)
+  fun write(source: ByteArray): BufferedSink
+
+  /**
+   * Like [OutputStream.write], this writes `byteCount`
+   * bytes of `source`, starting at `offset`.
+   */
+  @Throws(IOException::class)
+  fun write(source: ByteArray, offset: Int, byteCount: Int): BufferedSink
```

Comment:
for follow-up, I’d love to give `offset` and `byteCount` defaults and combine this with the preceding method

Reply:
(and we’d default `byteCount` to `source.length - offset` !)